### PR TITLE
[SDK-2594] Add helper function for silently retrieving session state without invoking authentication flow

### PR DESCRIPTION
**Pull request by evansims**
As I've been working to update the quickstarts for 8.0 I've noticed an area for improvement: currently we recommend invoking the getUser(), getIdToken(), etc. functions to retrieve the session state. However, all of these functions inherently begin an authentication flow when the requested data is not available. Further, if they encounter an error during this flow, they raise an exception, which creates extra work for customers who simply want to sniff for an available session presence and avoid creating error handling code around every instance of these functions.

I'm adding a function, getState(), which will return an object representing the stored session data, if it is available, for the current request. If none is available, it will simply return null. This will allow the developer to more easily display a login prompt within their application, or render content in a "guest mode" without writing extra error handling or writing work arounds to avoid beginning an authentication flow. It will simplify real world developer life/usage with the library quite a bit.

Usage:

```php
use Auth0\SDK\Auth0;

$sdk = new Auth0(/* ... configuration */);

// If we've just returned from the callback, remove the ?code parameter from the query.
if ($sdk->getRequestParameter('code')) {
  header("Location: /");
  exit;
}

// Use the new helper to silently get state.
$state = $sdk->getState();

// If there's no session, begin authentication flow. Alternatively customer could render their site in guest mode, or offer to login first, etc.
if (! $state) {
   $sdk->login();
}

// We have a session available.
if ($state) {
  // If the access token has expired, try to renew it.
  if ($state->accessTokenExpired) {
    try {
      $sdk->renew();
    } catch (\Auth0\SDK\Exception\StateException $exception) {
      // Couldn't renew the token, we might not have one based on our requested scopes. Let's fallback to starting a fresh authentication flow.
      $sdk->login();
    }
  }

  // 👌 Everything is good. Let's just dump info about the user, as an example.
  var_dump($state->user);
}

echo "Done!";
```

This is designed to give developers writing more robust applications who need more control a bit more flexibility without them having to write a bunch of custom code themselves.

### CODE REVIEW

File: src/Auth0.php [hunk 1]
```diff
@@ -245,8 +245,11 @@ public function exchange(): bool
         $user = $this->state->getIdTokenDecoded();
 
         if ($user === null || $this->configuration->getQueryUserInfo() === true) {
-            $user = $this->authentication()->userInfo($this->state->getAccessToken());
-            $user = HttpResponse::decodeContent($user);
+            $response = $this->authentication()->userInfo($response['access_token']);
+
+            if (HttpResponse::wasSuccessful($response)) {
+                $user = HttpResponse::decodeContent($response);
+            }
```
Comment (evansims) [inline src/Auth0.php:252]: Unrelated to this PR really, but had forgotten to do a success check here before attempting to decode the content, so sneaking that in here.

File: src/Auth0.php [hunk 2]
```diff
@@ -291,6 +294,37 @@ public function renew(
...
+            return null;
+        }
+
+        $token = $this->state->getIdToken();
+        $accessToken = $this->state->getAccessToken();
+        $accessTokenExpiration = (int) $this->state->getAccessTokenExpiration();
+        $accessTokenExpired = false;
+        $refreshToken = $this->state->getRefreshToken();
+
+        if (time() >= $accessTokenExpiration) {
```
Comment (lbalmaceda) [inline src/Auth0.php:314]: would it make sense to replace

`$accessTokenExpired = false;` with `$accessTokenExpired = time() >= $accessTokenExpiration;` ?

File: src/Auth0.php [hunk 3]
```diff
@@ -291,6 +294,37 @@ public function renew(
...
+            $accessTokenExpired = true;
+        }
+
+        return (object) [
+            'user' => $user,
+            'token' => $token,
```
Comment (lbalmaceda) [inline src/Auth0.php:320]: I'm confused by `token` vs `accessToken`. Could we use `idToken` if that's what the first one is, instead?

File: src/Auth0.php [hunk 4]
```diff
@@ -291,6 +294,37 @@ public function renew(
...
+            'accessToken' => $accessToken,
+            'accessTokenExpiration' => $accessTokenExpiration,
```
Comment (lbalmaceda) [inline src/Auth0.php:322]: is this a date or the lifetime of the token? What's the use case for consulting this vs checking the boolean prop?

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

File: src/Auth0.php [hunk 5]
```diff
@@ -291,6 +294,37 @@ public function renew(
         }
     }
 
+    /**
+     * Return an object representing the current session state (including id token, access token, access token expiration, refresh token and user data) without triggering an authorization flow. Returns null when session data is not available.
+     */
+    public function getState(): ?object
```
Comment (lbalmaceda) [inline src/Auth0.php:300]: I fear "state" could be confused with the one we send over for authentication (i.e. state & nonce). Might be better to use a different name. In Android and Swift we call them [Credentials](https://github.com/auth0/Auth0.swift/blob/master/Auth0/Credentials.swift) for example

Comment (evansims) [reply src/Auth0.php:300]: That's a good call out, I think that offers better clarity too. I'll update the method name.

File: src/Auth0.php (hunk 4 above)
Comment (evansims) [reply src/Auth0.php:322]: accessTokenExpiration is a unix timestamp representing the lifetime of the token; it's the time of the code exchange + the "expires_in" response from the API during code exchange.

File: src/Auth0.php (hunk 3 above)
Comment (evansims) [reply src/Auth0.php:320]: Yup, I think that makes sense too. Updating

File: src/Auth0.php (hunk 2 above)
Comment (evansims) [reply src/Auth0.php:314]: Yeah, I like that, I'll update

File: src/Auth0.php [hunk 6]
```diff
@@ -291,6 +294,33 @@ public function renew(
...
+     */
+    public function getCredentials(): ?object
+    {
+        $user = $this->state->getUser();
+
+        if (! $user) {
+            return null;
+        }
+
+        $token = $this->state->getIdToken();
```
Comment (lbalmaceda) [inline src/Auth0.php:308]: nit: could we also rename this method-scoped variable to `idToken`?

**Review by lbalmaceda — CHANGES_REQUESTED** [review]

**Review by lbalmaceda — APPROVED** [review]

**Pull request merged**